File: common/security/src/main/java/com/example/chat/auth/jwt/error/AuthExceptions.java
package com.example.chat.auth.jwt.error;

import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.server.resource.BearerTokenError;
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes;

/**
 * AuthException 생성 팩토리
 */
public final class AuthExceptions {

    private static final String EXPIRED_INDICATOR = "expired";

    private AuthExceptions() {
    }

    public static JwtAuthenticationException of(AuthErrorCode errorCode) {
        return new JwtAuthenticationException(errorCode);
    }

    public static JwtAuthenticationException fromBearerTokenError(BearerTokenError error) {
        String code = error.getErrorCode();
        String desc = error.getDescription();

        if (BearerTokenErrorCodes.INSUFFICIENT_SCOPE.equals(code)) {
            return new JwtAuthenticationException(AuthErrorCode.INSUFFICIENT_SCOPE);
        }

        if (BearerTokenErrorCodes.INVALID_TOKEN.equals(code)) {
            AuthErrorCode authErrorCode = isExpired(desc)
                    ? AuthErrorCode.EXPIRED_TOKEN
                    : AuthErrorCode.INVALID_TOKEN;
            return new JwtAuthenticationException(authErrorCode);
        }

        return new JwtAuthenticationException(AuthErrorCode.INVALID_TOKEN);
    }

    public static JwtAuthenticationException fromOAuth2Exception(OAuth2AuthenticationException ex) {
        if (ex.getError() instanceof BearerTokenError bearerTokenError) {
            return fromBearerTokenError(bearerTokenError);
        }
        return new JwtAuthenticationException(AuthErrorCode.INVALID_TOKEN);
    }

    public static boolean isExpired(String description) {
        return description != null && description.toLowerCase().contains(EXPIRED_INDICATOR);
    }
}

File: common/security/src/main/kotlin/com/example/chat/auth/jwt/error/AuthExceptions.kt

package com.example.chat.auth.jwt.error

import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.oauth2.server.resource.BearerTokenError
import org.springframework.security.oauth2.server.resource.BearerTokenErrorCodes

/**
 * AuthException 생성 헬퍼 유틸
 */
object AuthExceptions {
    const val EXPIRED_INDICATOR = "expired"

    fun of(errorCode: AuthErrorCode): JwtAuthenticationException {
        return JwtAuthenticationException(errorCode)
    }

    fun fromBearerTokenError(error: BearerTokenError): JwtAuthenticationException {
        val code = error.errorCode
        val desc = error.description

        if (BearerTokenErrorCodes.INSUFFICIENT_SCOPE == code) {
            return JwtAuthenticationException(AuthErrorCode.INSUFFICIENT_SCOPE)
        }

        if (BearerTokenErrorCodes.INVALID_TOKEN == code) {
            val expired = isExpired(desc)
            return JwtAuthenticationException(
                if (expired) {
                    AuthErrorCode.EXPIRED_TOKEN
                } else {
                    AuthErrorCode.INVALID_TOKEN
                }
            )
        }

        return JwtAuthenticationException(AuthErrorCode.INVALID_TOKEN)
    }

    fun fromOAuth2Exception(ex: OAuth2AuthenticationException): JwtAuthenticationException {
        val error = ex.error
        if (error is BearerTokenError) {
            return fromBearerTokenError(error)
        }
        return JwtAuthenticationException(AuthErrorCode.INVALID_TOKEN)
    }

    fun isExpired(description: String?): Boolean {
        return description != null && description.lowercase().contains(EXPIRED_INDICATOR)
    }
}
